### PR TITLE
 [FT-681] Transactional inconsistency with rollback

### PR DESCRIPTION
**Pull request by Jun-Yuan**
The indent and refactoring of funcs. re-commited and re-request the pull
http://jenkins.percona.com/view/TokuDB/job/PerconaFT-param/36/

### CODE REVIEW

**Comment by christianrober** [comment]
Looks good to me.

**Comment by georgelorchpercona** [comment]
http://jenkins.percona.com/view/TokuDB/job/PerconaFT-param/36/

Branch was renamed since jenkins, was FT-691 when it should have been FT-681. Commit is different due to some code formatting difffs from jenkins run and additional comments added to test case.

File: src/tests/rollback-inconsistency.cc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+populate_table(int start, int end, DB_TXN * parent, DB_ENV * env, DB * db) {
+    DB_TXN *txn = NULL;
+    int r = env->txn_begin(env, parent, &txn, 0); assert_zero(r);
+    for (int i = start; i < end; i++) {
+        int k = htonl(i);
+        char kk[4]; 
+        char str[220];
+        memset(kk, 0, sizeof kk);
+        memcpy(kk, &k, sizeof k);
+        memset(str,'a', sizeof str-1);
```
Comment (BohuTANG) [inline src/tests/rollback-inconsistency.cc]: should memset(str, 0, sizeof str) first, or we will get uninitialized error.

Comment (kuszmaul) [inline src/tests/rollback-inconsistency.cc]: This code is kind of unclean.  Better would be

 int k = htonl(i);
 char str[220];
 memset(str, 'a', sizeof str);
 DBT key = { .data = k, .size = sizeof k};
 DBT val = { .data = str, .size = sizeof str};

1) Don't create a separate character array when k is just fine.  (You might have to cast k to an appropriate pointer when initialing the key dbt).
 2) Don't repeat the literal value "220" in the initializion of val.
 3) Don't memset str twice.  Just set the last entry to 0 if you want a 0-terminated string.  It doesn't look like this test requires a 0-terminates string, however.  In fact, the test doesn't really rely on the value of the string being any particular value. So unless valgrind is complaining, the test is arguably OK with uninitialized str[219].  Valgrind will likely complain, however, when the database is written to disk.

Comment (Jun-Yuan) [inline src/tests/rollback-inconsistency.cc]: Hi, Thanks @BohuTANG and @kuszmaul for pointing out problems. I actually copied the part of key and value from other code ... I should have been more careful with it.  Will correct these issue together with my next pull request for fixing read-commit bug. Thanks.

Comment (kuszmaul) [inline src/tests/rollback-inconsistency.cc]: Yes, some of those tests (especially the old ones) were messy...

-Bradley

**Comment by BohuTANG** [comment]
So we should change valgrind suites to thread/addresssanitizer, it's faster and for all tests.

**Comment by prohaska7** [comment]
I would add address sanitizer and thread sanitizer to tests rather than
replace valgrind/helgrind/drd with address/thread sanitizer.  There is
still innovation in the valgrind toolset that we probably want to work with.

The use of the thread sanitizer with the Tokutek code is a bit of a project
since there are lots of deliberately racy variables that are marked with
helgrind annotations so that data races on them are suppressed.

I did an experiment running the tokubackup tests with the thread
sanitizer.  Results are here:
http://prohaska7.blogspot.com/2015/10/tokubackup-versus-sanitizers.html.
 gcov found code uncovered by tests.  Address sanitizer found a memory
leak.  Thread sanitizer found a real problem in tokubackup.

Since the FT software is much larger than tokubackup, converting it to run
with the thread sanitizer will be a larger project.  The main problem is
that helgrind suppressions of memory ranges at run time are not supported
by the thread sanitizer.  So, the method that FT used to annotate racy
variables needs to be replaced.  IMO, all racy variables should use either
be c++ atomic variables OR use explicit functions to load/store so that
thread sanitizer suppressions can be added.

**Pull request merged**